### PR TITLE
feat(whale-api): add max hops in path finding

### PR DESCRIPTION
**Pull request by pierregee**
#### What this PR does / why we need it:
- 3 max hops is a consensus restriction at the moment, to prevent costly mistakes or accidental loops. This PR is to ignore paths that are > 3 hops

#### Which issue(s) does this PR fixes?:

Fixes #

#### Additional comments?:
- Most code changes are only to adjust the ids because of the added tokens
- Added two tokens (M, N) to be able to test > 3 hops
```mermaid
graph TD;
    I---J;
    J---L;
    J---K;
    L---K;
    L---M*;
    M*---N*;
```

### CODE REVIEW

**Review by fuxingloh — DISMISSED** [review]
lgtm

File: apps/whale-api/src/module.api/poolswap.pathfinding.service.ts [hunk 1]
```diff
@@ -135,6 +135,10 @@ export class PoolSwapPathFindingService {
     const poolPairPaths: SwapPathPoolPair[][] = []
 
     for (const path of allSimplePaths(this.tokenGraph, fromTokenId, toTokenId)) {
+      if (path.length > 4) {
```
Comment (eli-lim) [inline apps/whale-api/src/module.api/poolswap.pathfinding.service.ts:139]: Would add a quick explanation of this magic number here

```suggestion
      // Max 3 hops, e.g. A --> B --> C --> D
      if (path.length > 4) {
```

File: apps/whale-api/src/module.api/poolpair.controller.e2e.ts [hunk 2]
```diff
@@ -452,6 +469,29 @@ describe('get best path', () => {
     })
   })
 
+  it('should ignore correct swap path > 3 legs', async () => {
+    const response = await controller.getBestPath('9', '14') // I to N
+    /* paths available
+      (4 legs) I-J -> J-L -> L-M -> M-N
+      (5 legs) I-J -> J-K -> K-L -> L-M -> M-N
```
Comment (eli-lim) [inline apps/whale-api/src/module.api/poolpair.controller.e2e.ts]:
```suggestion
      (4 legs) Swap I through -> I-J -> J-L -> L-M -> M-N to get N
      (5 legs) Swap I through -> I-J -> J-K -> K-L -> L-M -> M-N to get N
```

File: apps/whale-api/src/module.api/poolpair.controller.e2e.ts [hunk 3]
```diff
@@ -612,6 +652,28 @@ describe('get all paths', () => {
     })
   })
 
+  it('should ignore correct swap paths > 3 legs', async () => {
+    const response = await controller.listPaths('9', '14') // I to N
+
+    /* paths available
+      (4 legs) I-J -> J-L -> L-M -> M-N
+      (5 legs) I-J -> J-K -> K-L -> L-M -> M-N
```
Comment (eli-lim) [inline apps/whale-api/src/module.api/poolpair.controller.e2e.ts]:
```suggestion
      (4 legs) Swap I through -> I-J -> J-L -> L-M -> M-N to get N
      (5 legs) Swap I through -> I-J -> J-K -> K-L -> L-M -> M-N to get N
```
```

**Review by eli-lim — COMMENTED** [review]
LGTM generally
Just some nits on adding comments!

File: apps/whale-api/src/module.api/poolpair.controller.e2e.ts (hunk 3 above)
Comment (pierregee) [reply apps/whale-api/src/module.api/poolpair.controller.e2e.ts]: Updated!

File: apps/whale-api/src/module.api/poolpair.controller.e2e.ts (hunk 2 above)
Comment (pierregee) [reply apps/whale-api/src/module.api/poolpair.controller.e2e.ts]: Updated!

File: apps/whale-api/src/module.api/poolswap.pathfinding.service.ts (hunk 1 above)
Comment (pierregee) [reply apps/whale-api/src/module.api/poolswap.pathfinding.service.ts:139]: Updated!

File: apps/whale-api/src/module.api/poolpair.controller.e2e.ts [hunk 4]
```diff
@@ -965,14 +1043,14 @@ describe('latest dex prices', () => {
   })
 
   it('should list DAT tokens only - M (non-DAT token) is not included in result', async () => {
```
Comment (eli-lim) [inline apps/whale-api/src/module.api/poolpair.controller.e2e.ts]:
```suggestion
  it('should list DAT tokens only - O (non-DAT token) is not included in result', async () => {
```

Comment (eli-lim) [reply apps/whale-api/src/module.api/poolpair.controller.e2e.ts]: ONE LAST `nit` 🤣

Comment (pierregee) [reply apps/whale-api/src/module.api/poolpair.controller.e2e.ts]: oh 💩 I forgot this one!

Comment (pierregee) [reply apps/whale-api/src/module.api/poolpair.controller.e2e.ts]: Updated!

File: apps/whale-api/src/module.api/poolswap.pathfinding.service.ts [hunk 5]
```diff
@@ -135,6 +135,11 @@ export class PoolSwapPathFindingService {
     const poolPairPaths: SwapPathPoolPair[][] = []
 
     for (const path of allSimplePaths(this.tokenGraph, fromTokenId, toTokenId)) {
+      // Max 3 hops, e.g. A --> B --> C --> D
+      if (path.length > 4) {
```
Comment (canonbrother) [inline apps/whale-api/src/module.api/poolswap.pathfinding.service.ts:139]: heyy.. need clarify.. so the the `I-J` won't be in the path right??
```
I-J -> J-L -> L-M -> M-N
// to
I-J -> path(J-L -> L-M -> M-N)
```

Comment (pierregee) [reply apps/whale-api/src/module.api/poolswap.pathfinding.service.ts:139]: Ah in this example, `I-J -> J-L -> L-M -> M-N` is already 4 hops, so `I-J` is in the path
path.length = `[I, J, L, M, N]`. Hence path.length > 4 will be ignored

**Review by canonbrother — APPROVED** [review]

**Review by eli-lim — APPROVED** [review]
:shipit:

**Review by fuxingloh — APPROVED** [review]

**Pull request merged**